src/filter.rs

```diff
@@ -0,0 +1,322 @@
+//! Interfaces for adding custom filter to libgit2
+
+use libc::{c_int, c_char, c_void, size_t};
+use std::{mem, slice, ptr};
+use {raw, call, util, Error, panic};
+use std::ffi::CString;
+use std::path::Path;
+use raw::{git_filter, git_writestream};
+
+#[repr(C)]
+struct Filter {
+    filter: git_filter,
+    attrs: Option<CString>,
+    instance: Box<StreamFilter>,
+}
+
+#[repr(C)]
+struct Stream {
+    parent: git_writestream,
+    handler: Box<WriteStream>,
+}
+
+/// Wrapper of the raw git_writestream object, representing the input
+/// of next filter of the chain
+pub struct RawWriteStream {
+    stream: *mut git_writestream,
+}
+
+/// A trait representing operations to handle filter data stream.
+///
+/// These callbacks are typically called by libgit2.
+pub trait WriteStream : Send + 'static {
+    /// New stream data, Implementations should process the buffer data
+    /// and then write the output to the next stream if needed.
+    fn write(&mut self, buf: &[u8]) -> Result<(), Error>;
+
+    /// Close stream. Implementations should do finalizing work and then
+    /// close the next stream.
+    fn close(&mut self) -> Result<(), Error>;
+}
+
+/// Information to construct a new stream filter.
+pub struct FilterSource {
+    raw: *const raw::git_filter_source
+}
+
+impl FilterSource {
+    fn from_raw(raw: *const raw::git_filter_source) -> FilterSource {
+        FilterSource {
+            raw: raw
+        }
+    }
+
+    /// Returns `true` if it is a smudge filter, else a clean filter.
+    pub fn to_worktree(&self) -> bool {
+        unsafe {
+            raw::git_filter_source_mode(self.raw) == raw::GIT_FILTER_TO_WORKTREE
+        }
+    }
+
+    /// Returns the path that the source data is coming from.
+    pub fn source_path(&self) -> &Path {
+        unsafe {
+            let ptr = raw::git_filter_source_path(self.raw);
+            util::bytes2path(::opt_bytes(self, ptr).unwrap())
+        }
+    }
+}
+
+/// A trait representing a customized filter to be registered into libgit2
+/// filter chain.
+///
+/// Modeled after git2/sys/filter.h in a streaming manner.
+pub trait StreamFilter : Send + Sync + 'static {
+	/// Called when the filter is first used for any file.
+    fn init(&self) -> Result<(), Error> { Ok(()) }
+
+    /// Create a new stream to handle new blob, should return a initialized
+    /// WriteStream.
+    ///
+    /// Custom filters can retrieve information of source blob from `source`,
+    /// and custom filter output should be written to `next` later.
+    fn open_stream(&self, source: FilterSource, next: RawWriteStream)
```

Comment:
Also maybe this should pass in `&FilterSource` as it's not passing ownership down?